src/boot/efi/linux.c

```diff
@@ -15,9 +15,61 @@
 #include "linux.h"
 #include "missing_efi.h"
 #include "pe.h"
+#include "secure-boot.h"
 #include "util.h"
 
-EFI_STATUS load_trusted_image(EFI_HANDLE parent, const void *source, size_t len, EFI_HANDLE *ret_image) {
+static struct TrustedState {
+        EFI_SECURITY_FILE_AUTHENTICATION_STATE orig_auth;
+        EFI_SECURITY2_FILE_AUTHENTICATION orig_auth2;
+        const void *buffer;
+        size_t size;
+        const EFI_DEVICE_PATH *device_path;
+} *trusted = NULL;
+
+static EFIAPI EFI_STATUS security_hook(
+                const EFI_SECURITY_ARCH_PROTOCOL *this,
+                uint32_t authentication_status,
+                const EFI_DEVICE_PATH *file) {
+
+        EFI_STATUS err;
+
+        assert(trusted);
+        assert(trusted->orig_auth);
+
+        err = trusted->orig_auth(this, authentication_status, file);
+        if (err == EFI_SUCCESS)
+                return EFI_SUCCESS;
+
+        if (file == trusted->device_path)
+                return EFI_SUCCESS;
```

Comment:
hmm, so why call into orig_auth() first here? tricking out shim?

Reply:
wouldn't it be better to compare device paths by contents instead of pointer? You never know what the auth layer might copy around internally...

Reply:
I'd rather only do that if it ever comes to that.

Reply:
Shim has nothing to do with this. It simply makes sense to call the original thingy in something you hook into.

Reply:
I wonder about that. I mean this stuff will show up in the audit logs, and that will be very confusing at least, and people might wonder how things failed there and we booted up anyway.

the function is not without side-effect, if we want to skip it we should really skip it, not let it run, and then eat up the result I am sure.